File: rust/2016/src/day04.rs
use aoc::iter::CollectArrayVec;
use itertools::Itertools;

struct Room {
    name: String,
    sector: usize,
    checksum: [char; 5],
}

// This char chosen because it appears before 'a'
const SEPARATOR: char = '`';

type Input = Vec<Room>;

fn parse(filename: &str) -> Input {
    let sep_str = format!("{}", SEPARATOR);
    aoc::file::read_lines(filename)
        .iter()
        .map(|line| {
            let parts = line.split('-').collect::<Vec<_>>();
            let num_parts = parts.len();
            let name = parts.iter().take(num_parts - 1).join(&sep_str);
            let tail = parts
                .iter()
                .skip(num_parts - 1)
                .next()
                .unwrap()
                .split_once('[')
                .unwrap();
            Room {
                name: name,
                sector: tail.0.parse().unwrap(),
                checksum: (&tail.1[..(tail.1.len() - 1)])
                    .chars()
                    .collect_array_vec::<5>()
                    .into_inner()
                    .unwrap(),
            }
        })
        .collect()
}

const NUM_LETTERS: usize = 26;

fn solve_case1(rooms: &Input) -> usize {
    rooms
        .iter()
        .filter_map(|room| {
            let mut frequency_map = <[usize; NUM_LETTERS]>::default();
            for letter in room.name.bytes() {
                if letter == (SEPARATOR as u8) {
                    continue;
                }
                frequency_map[(letter - b'a') as usize] += 1;
            }
            let mut sorted = frequency_map
                .iter()
                .enumerate()
                .collect_array_vec::<NUM_LETTERS>();
            sorted.sort_by(|(lhs_id, lhs), (rhs_id, rhs)| rhs.cmp(lhs).then(lhs_id.cmp(rhs_id)));
            sorted
                .into_iter()
                .zip(room.checksum.iter())
                .all(|((letter_id, _), check)| *check == ((letter_id as u8 + b'a') as char))
                .then_some(room.sector)
        })
        .sum()
}

fn solve_case2(rooms: &Input) -> usize {
    for room in rooms {
        let real_name = room
            .name
            .bytes()
            .map(|letter| {
                if letter == (SEPARATOR as u8) {
                    ' '
                } else {
                    ((((letter - b'a') as usize + room.sector) % NUM_LETTERS) as u8 + b'a') as char
                }
            })
            .collect::<String>();
        if real_name == "northpole object storage" {
            return room.sector;
        }
    }
    0
}

fn main() {
    println!("Part 1");
    let example = parse("day04.example");
    aoc::expect_result!(1514, solve_case1(&example));
    let input = parse("day04.input");
    aoc::expect_result!(173787, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(0, solve_case2(&example));
    aoc::expect_result!(548, solve_case2(&input));
}

File: 2016/day04.cpp
// https://adventofcode.com/2016/day/4

#include "../common/common.h"
#include "../common/rust.h"

#ifndef AOC_IMPORT_STD
#include <algorithm>
#include <array>
#include <numeric>
#include <print>
#endif

// The name keeps the dashes that separate its words
constexpr let SEPARATOR = '-';

constexpr let NUM_LETTERS = 26uz;
constexpr let CHECKSUM_LENGTH = 5uz;

struct Room {
  String name;
  usize sector;
  std::array<char, CHECKSUM_LENGTH> checksum;
};

using Input = Vec<Room>;

auto parse(String const& filename) -> Input {
  return aoc::views::read_lines(filename) |
         stdv::transform([](str line) {
           let tail_start = line.rfind(SEPARATOR) + 1;
           let[sector, checksum] =
               aoc::split_once(line.substr(tail_start), '[');
           auto room = Room{
               .name = String{line.substr(0, tail_start - 1)},
               .sector = aoc::to_number<usize>(sector),
               .checksum = {},
           };
           // The checksum still carries the closing bracket
           stdr::copy_n(checksum.begin(), CHECKSUM_LENGTH,
                        room.checksum.begin());
           return room;
         }) |
         aoc::collect_vec<Room>();
}

fn solve_case1(Input const& rooms) -> usize {
  auto sum = usize{};
  for (let& room : rooms) {
    auto frequency = std::array<usize, NUM_LETTERS>{};
    for (let letter : room.name) {
      if (letter != SEPARATOR) {
        ++frequency[static_cast<usize>(letter - 'a')];
      }
    }
    auto sorted_ids = aoc::views::indices_of(frequency) |
                      aoc::collect_static_vec<u8, NUM_LETTERS>();
    stdr::sort(sorted_ids, [&](u8 lhs, u8 rhs) {
      return (frequency[lhs] != frequency[rhs])
                 ? (frequency[lhs] > frequency[rhs])
                 : (lhs < rhs);
    });
    if (stdr::equal(       //
            room.checksum, //
            sorted_ids |
                stdv::take(CHECKSUM_LENGTH) |
                stdv::transform([](u8 letter_id) {
                  return static_cast<char>(letter_id + 'a');
                }))) {
      sum += room.sector;
    }
  }
  return sum;
}

fn solve_case2(Input const& rooms) -> usize {
  for (let& room : rooms) {
    let real_name =
        room.name |
        stdv::transform([&](char letter) {
          if (letter == SEPARATOR) {
            return ' ';
          }
          let rotated =
              (static_cast<usize>(letter - 'a') + room.sector) % NUM_LETTERS;
          return static_cast<char>(rotated + 'a');
        }) |
        aoc::collect_string();
    if (real_name == "northpole object storage") {
      return room.sector;
    }
  }
  return 0;
}

int main() {
  std::println("Part 1");
  let example = parse("day04.example");
  AOC_EXPECT_RESULT(1514, solve_case1(example));
  let input = parse("day04.input");
  AOC_EXPECT_RESULT(173787, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(0, solve_case2(example));
  AOC_EXPECT_RESULT(548, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}
